iob_soc.py

```diff
@@ -40,308 +43,606 @@ class iob_soc(iob_module):
-            "if_defined": "USE_EXTMEM",
-            "ports": [],
-        },
-    ]
-
-    # IOb-SoC has the following set of non standard attributes:
-    peripherals = [
-        uart.instance("UART0"),
-        timer.instance("TIMER0"),
-    ]
-    peripheral_portmap = [  # List of tuples, each tuple corresponds to a port map
-        (
-            {
-                "corename": "UART0",
-                "if_name": "rs232",
-                "port": "txd",
-                "bits": [],
-            },
-            {
-                "corename": "external",
-                "if_name": "uart",
-                "port": "uart_txd_o",
-                "bits": [],
-            },
+            descr="Bus of AXI master interfaces for external memory. One interface for this system and others optionally for peripherals.",
+            if_defined="USE_EXTMEM",
+            ports=[],
+            connect_to=axi
         ),
-        (
-            {
-                "corename": "UART0",
-                "if_name": "rs232",
-                "port": "rxd",
-                "bits": [],
-            },
-            {
-                "corename": "external",
-                "if_name": "uart",
-                "port": "uart_rxd_i",
-                "bits": [],
-            },
+        iob_interface(
+            name="rs232",
+            type="master",
+            port_prefix="",
+            wire_prefix="",
+            descr="iob-soc uart interface",
+            ports=[
+                iob_port(
+                    name="txd",
+                    direction="output",
+                    width=1,
+                    descr="UART transmit data",
+                    connect_to=txd,
+                ),
+                iob_port(
+                    name="rxd",
+                    direction="input",
+                    width=1,
+                    descr="UART receive data",
+                    connect_to=rxd,
+                ),
+                iob_port(
+                    name="rts",
+                    direction="output",
+                    width=1,
+                    descr="UART request to send",
+                    connect_to=rts,
+                ),
+                iob_port(
+                    name="cts",
+                    direction="input",
+                    width=1,
+                    descr="UART clear to send",
+                    connect_to=cts,
+                ),
+            ],
         ),
-        (
-            {
-                "corename": "UART0",
-                "if_name": "rs232",
-                "port": "cts",
-                "bits": [],
+    ]
+
+    #######################################
+    # IOb-SoC modules, wires, and instances
+    #######################################
+
+    #
+    # SYSTEM RESET
+    #
+
+    boot = iob_wire("boot", width=1)
+    cpu_reset = iob_wire("cpu_reset", width=1)
+
+    #
+    # CPU
+    #
+
+    cpu_i_req = iob_wire("cpu_i_req", width=REQ_W)
```

Comment:
Currently, the `REQ_W` macro does not exist in python.
We could convert this to a string and let the generator insert it as is in the generated Verilog. This would allow us to keep using the `` `REQ_W`` macro in verilog.

Reply:
we have no interest in REQ_W

Reply:
Currently, this PR only shows a translation from the current `iob_soc.v` implemented in Verilog, to the `py2hw` syntax in Python.
Removing the `REQ_W`, `RESP_W`, and associated macros would probably mean a major restructure of the current iob_soc module implementation.
If the objective is to keep this PR simple to understand, then this change should probably be done in another PR.